{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "fac11c3e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Image saved: Gillespie_traces/Gillespie_trace249.png\n"
     ]
    },
    {
     "ename": "FileNotFoundError",
     "evalue": "No such file: 'C:\\Users\\saman\\Gillespie_traces\\Gillespie_trace0.png'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[17], line 94\u001b[0m\n\u001b[0;32m     92\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m filename \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mrange\u001b[39m(num_time_step):\n\u001b[0;32m     93\u001b[0m     image_path \u001b[38;5;241m=\u001b[39m os\u001b[38;5;241m.\u001b[39mpath\u001b[38;5;241m.\u001b[39mjoin(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mGillespie_traces\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mGillespie_trace\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mfilename\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m.png\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m---> 94\u001b[0m     image \u001b[38;5;241m=\u001b[39m \u001b[43mimageio\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mimread\u001b[49m\u001b[43m(\u001b[49m\u001b[43mimage_path\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m     95\u001b[0m     writer\u001b[38;5;241m.\u001b[39mappend_data(image)\n",
      "File \u001b[1;32m~\\anaconda3\\lib\\site-packages\\imageio\\v2.py:226\u001b[0m, in \u001b[0;36mimread\u001b[1;34m(uri, format, **kwargs)\u001b[0m\n\u001b[0;32m    223\u001b[0m imopen_args \u001b[38;5;241m=\u001b[39m decypher_format_arg(\u001b[38;5;28mformat\u001b[39m)\n\u001b[0;32m    224\u001b[0m imopen_args[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mlegacy_mode\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n\u001b[1;32m--> 226\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m imopen(uri, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mri\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mimopen_args) \u001b[38;5;28;01mas\u001b[39;00m file:\n\u001b[0;32m    227\u001b[0m     result \u001b[38;5;241m=\u001b[39m file\u001b[38;5;241m.\u001b[39mread(index\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m0\u001b[39m, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n\u001b[0;32m    229\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m result\n",
      "File \u001b[1;32m~\\anaconda3\\lib\\site-packages\\imageio\\core\\imopen.py:113\u001b[0m, in \u001b[0;36mimopen\u001b[1;34m(uri, io_mode, plugin, extension, format_hint, legacy_mode, **kwargs)\u001b[0m\n\u001b[0;32m    111\u001b[0m     request\u001b[38;5;241m.\u001b[39mformat_hint \u001b[38;5;241m=\u001b[39m format_hint\n\u001b[0;32m    112\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m--> 113\u001b[0m     request \u001b[38;5;241m=\u001b[39m \u001b[43mRequest\u001b[49m\u001b[43m(\u001b[49m\u001b[43muri\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mio_mode\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mformat_hint\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mformat_hint\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mextension\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mextension\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    115\u001b[0m source \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m<bytes>\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(uri, \u001b[38;5;28mbytes\u001b[39m) \u001b[38;5;28;01melse\u001b[39;00m uri\n\u001b[0;32m    117\u001b[0m \u001b[38;5;66;03m# fast-path based on plugin\u001b[39;00m\n\u001b[0;32m    118\u001b[0m \u001b[38;5;66;03m# (except in legacy mode)\u001b[39;00m\n",
      "File \u001b[1;32m~\\anaconda3\\lib\\site-packages\\imageio\\core\\request.py:247\u001b[0m, in \u001b[0;36mRequest.__init__\u001b[1;34m(self, uri, mode, extension, format_hint, **kwargs)\u001b[0m\n\u001b[0;32m    244\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mInvalid Request.Mode: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mmode\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m    246\u001b[0m \u001b[38;5;66;03m# Parse what was given\u001b[39;00m\n\u001b[1;32m--> 247\u001b[0m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_parse_uri\u001b[49m\u001b[43m(\u001b[49m\u001b[43muri\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    249\u001b[0m \u001b[38;5;66;03m# Set extension\u001b[39;00m\n\u001b[0;32m    250\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m extension \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n",
      "File \u001b[1;32m~\\anaconda3\\lib\\site-packages\\imageio\\core\\request.py:407\u001b[0m, in \u001b[0;36mRequest._parse_uri\u001b[1;34m(self, uri)\u001b[0m\n\u001b[0;32m    404\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m is_read_request:\n\u001b[0;32m    405\u001b[0m     \u001b[38;5;66;03m# Reading: check that the file exists (but is allowed a dir)\u001b[39;00m\n\u001b[0;32m    406\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m os\u001b[38;5;241m.\u001b[39mpath\u001b[38;5;241m.\u001b[39mexists(fn):\n\u001b[1;32m--> 407\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mFileNotFoundError\u001b[39;00m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mNo such file: \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;241m%\u001b[39m fn)\n\u001b[0;32m    408\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m    409\u001b[0m     \u001b[38;5;66;03m# Writing: check that the directory to write to does exist\u001b[39;00m\n\u001b[0;32m    410\u001b[0m     dn \u001b[38;5;241m=\u001b[39m os\u001b[38;5;241m.\u001b[39mpath\u001b[38;5;241m.\u001b[39mdirname(fn)\n",
      "\u001b[1;31mFileNotFoundError\u001b[0m: No such file: 'C:\\Users\\saman\\Gillespie_traces\\Gillespie_trace0.png'"
     ]
    }
   ],
   "source": [
    "import math\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import imageio.v2 as imageio\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "num_substrate = 20\n",
    "num_enzyme = 10\n",
    "num_trace = 100\n",
    "\n",
    "num_time_step = 250\n",
    "simulation_time = 2.5\n",
    "\n",
    "dt = simulation_time / num_time_step\n",
    "q1 = num_enzyme\n",
    "\n",
    "c1 = 1.2\n",
    "c2 = 0.3\n",
    "c3 = 2.9\n",
    "\n",
    "v1 = [-1, -1]\n",
    "v2 = [1, 1]\n",
    "v3 = [0, 1]\n",
    "\n",
    "a1 = lambda enzyme, substrate: enzyme * substrate * c1\n",
    "a2 = lambda enzyme: (q1 - enzyme) * c2\n",
    "a3 = lambda enzyme: (q1 - enzyme) * c3\n",
    "\n",
    "trace_time = [0 for _ in range(num_trace)]\n",
    "trace_substrate_state = [num_substrate for _ in range(num_trace)]\n",
    "trace_enzyme_state = [num_enzyme for _ in range(num_trace)]\n",
    "\n",
    "\n",
    "def calc_p_from_trace(trace_substrate_state, trace_enzyme_state, num_substrate, num_enzyme, tSteps):\n",
    "    P = pd.DataFrame([[0 for _ in range(num_substrate)] for _ in range(num_enzyme)])\n",
    "    \n",
    "    #There was a problem with the for loop so I\n",
    "    #changed it and I'm not sure if it's the right way to do it but it works\n",
    "    for t in range(len(trace_substrate_state)):\n",
    "        substrate_idx = min(trace_substrate_state[t], num_substrate - 1)# this ensures that the index is within the bounds of the DataFrame's columns \n",
    "        enzyme_idx = min(trace_enzyme_state[t], num_enzyme - 1) # this ensures that the index is within the bounds of the DataFrame's rows\n",
    "        P.iloc[enzyme_idx, substrate_idx] += 1\n",
    "     \n",
    "    \n",
    "    P = P / num_trace\n",
    "    sns.heatmap(P, annot=True)\n",
    "    plt.xlabel('Number of Substrates') \n",
    "    plt.ylabel('Number of Enzyme')\n",
    "    plt.savefig(f\"Gillespie_traces/Gillespie_trace{tSteps}.png\")\n",
    "    print(f\"Image saved: Gillespie_traces/Gillespie_trace{tSteps}.png\")\n",
    "    \n",
    "    plt.close()\n",
    "    \n",
    "    return P\n",
    "\n",
    "\n",
    "for tSteps in range(num_time_step): \n",
    "    for t in range(num_trace):\n",
    "        enzyme, substrate = trace_enzyme_state[t], trace_substrate_state[t]\n",
    "        \n",
    "        while trace_time[t] < dt * tSteps:\n",
    "            a1X = a1(substrate, enzyme)\n",
    "            a2X = a2(enzyme)\n",
    "            a3X = a3(enzyme)\n",
    "            \n",
    "            rand_uniform = np.random.uniform(low=0.0, high=1.0, size=1)# why we generate a random number!?\n",
    "            xi1 = rand_uniform * (a1X + a2X + a3X)\n",
    "            \n",
    "            if (substrate, enzyme) == (0, 10):\n",
    "                trace_time[t] = simulation_time\n",
    "            elif xi1 < a1X:\n",
    "                substrate, enzyme = substrate + v1[0], enzyme + v1[1]\n",
    "            elif xi1 < a1X + a2X:\n",
    "                substrate, enzyme = substrate + v2[0], enzyme + v2[1]\n",
    "            else:\n",
    "                substrate, enzyme = substrate + v3[0], enzyme + v3[1]\n",
    "            \n",
    "            try:\n",
    "                xi2 = -math.log(1 - rand_uniform) / (a1X + a2X + a3X)# why this calculation?!\n",
    "            except ZeroDivisionError:\n",
    "                xi2 = 10000\n",
    "            \n",
    "            trace_time[t] += xi2\n",
    "        \n",
    "        trace_substrate_state[t] = substrate\n",
    "        trace_enzyme_state[t] = enzyme\n",
    "\n",
    "P = calc_p_from_trace(trace_substrate_state, trace_enzyme_state, num_substrate, num_enzyme, tSteps)\n",
    "with imageio.get_writer('Gillespie_gif.gif', mode='I') as writer:\n",
    "    for filename in range(num_time_step):\n",
    "        image_path = os.path.join('Gillespie_traces', f'Gillespie_trace{filename}.png')\n",
    "        image = imageio.imread(image_path)\n",
    "        writer.append_data(image)  \n",
    "        \n",
    "        \n",
    "\n",
    "\"\"\"\n",
    "I tried to simulate, but the folder that is used to store\n",
    "images has a problem, which unfortunately could not be solved.\n",
    "\"\"\"\n",
    "\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
